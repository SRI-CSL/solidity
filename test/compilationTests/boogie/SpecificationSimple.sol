pragma solidity ^0.4.23;

/**
 * @notice invariant x == y
 * @notice some random comment
 */
contract SpecificationSimple {
    uint x;
    uint y;

    constructor() public {
        x = y = 0;
    }

    function add(uint amount) public {
        x += amount;
        y += amount;
    }

    function add_incorrect(uint amount) public {
        x += amount;
        y -= amount; // ERROR
    }

    function add_incorrect_2(uint amount) public {
        x += amount;
        add(amount); // ERROR, invariant does not hold when calling function
        y += amount;
    }
}