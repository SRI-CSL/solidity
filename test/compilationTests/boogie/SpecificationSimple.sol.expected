Verifying SpecificationSimple::[constructor]: OK
Verifying SpecificationSimple::add: OK
Verifying SpecificationSimple::add_incorrect: ERROR
 - Source SpecificationSimple.sol, line 20, col 5: Invariant 'x == y' might not hold at end of function.
Verifying SpecificationSimple::add_incorrect_2: ERROR
 - Source SpecificationSimple.sol, line 27, col 9: Invariant 'x == y' might not hold when entering function.
Errors were found by the verifier.
